### PR TITLE
added form validation to text input

### LINKED ISSUE #6: 6. As a user, I want to be alerted when I’m entering an item that is the same as something already on my list so I can avoid duplicate items.
AC:

* Show an error message if the user tries to submit a new item that has the exact same name as an existing item
* Show an error message if the user tries to submit a new item that has the same name as an existing item, where capitalization has been normalized and punctuation has been removed
* The user’s original input should be what gets saved in the database

### PR DESCRIPTION
**Pull request by daniel-sj-hong**
## Description

- Takes the input from the user and checked to see if they're trying to add an item that's already in the list. 
- Used regex to check if there was any punctuation in user's input and if there was, we got rid of it so that we can compare if that item already exists.
- Lastly, used toLowerCase() to convert item names to the same case, so they can be checked for equality.
- If there's a match, we show an error message saying that the item already exists in the list.
- If an item is not already in the list, you should see a message that an item was successfully added.

## Related Issue

closes #6 

## Acceptance Criteria

- Show an error message if the user tries to submit a new item that has the exact same name as an existing item
- Show an error message if the user tries to submit a new item that has the same name as an existing item, where capitalization has been normalized and punctuation has been removed
- The user’s original input should be what gets saved in the database

## Type of Changes

|     | Type                       |
| --- | -------------------------- |
|    | :bug: Bug fix              |
|  ✓ | :sparkles: New feature     |
|    | :hammer: Refactoring       |
|  ✓  | :100: Add tests            |
|    | :link: Update dependencies |
|    | :scroll: Docs              |

## Updates
- Added form validation to the input box

### After

Screenshot (119)
Screenshot 2022-04-20 174711

## Testing Steps / QA Criteria

- Look at the current list and see what item is already there. If there's no items add in some items e.g. bread, apples, etc
- Go back to the input field and try to add an item that's already on the list
- The input field should immediately reset and clear after you click the button
- An error message should pop up saying that the item you're trying to add is already in the list.
- No duplicate item should be added to the list and you can go to the list view just to make sure
- If the item was not a match, it should get added to the list
- To further, test add in an item that's already in the list with random punctuation and casing e.g. "B.r.E.a-D", and it should still come out as a match for 'bread' because we're filtering out and accounting for random casing and punctuation. If it's not a match, whatever the user originally typed with the random punctuation and casing will get added to the list. 
- We also added a notification for when an item is successfully added.

### CODE REVIEW

**Review by sparsell — APPROVED** [review]
Tested by loading app and added several items to my list:
- used erroneous punctuation and mix of caps/lower case when attempting to add the same item more than once
- successfully saw the error message that the item had already been added
- successfully saw the original item as first typed "incorrectly"

**Review by k-schrattenholzer — APPROVED** [review]
I was able to pull this down - and validate that an error appeared when I attempted to add an item twice (even when it had special characters or punctuation). This looks great - I love the use of Regex to account for those non alphabetical characters!

File: src/views/AddItem/AddItem.jsx [hunk 1]
```diff
@@ -19,7 +32,48 @@ export default function AddItem({ token }) {
...
+  const validate = () => {
+    //get the value the user typed
+    //preserve that value
+    //get the whole list from firebase
+    //loop through and check each list item for a match with what the user type
+    let status = false;
+    data.forEach((item) => {
+      let editedItem = item.name;
+      //use regex to find and replace any punctuation with and empty string
+      let result = editedItem.replace(/[\W|_]/g, '');
```
Comment (mxmason) [inline src/views/AddItem/AddItem.jsx:62]: 🔍 Love the use of RegEx here!

File: src/views/AddItem/AddItem.jsx [hunk 2]
```diff
@@ -19,7 +32,48 @@ export default function AddItem({ token }) {
...
+    } else {
+      addItem();
+      setListItem({
+        name: '',
+        frequency: '7',
+      });
+      setError('Item added successfully!');
+      setTimeout(() => {
+        setError('');
+      }, 3000);
```
Comment (mxmason) [inline src/views/AddItem/AddItem.jsx]: 🔍 Great work on the feature, and thank you for doing the additional confirmation message! One thing: there are some accessibility risks to getting rid of a notification on a timeout. What if a user hasn't read the notification before it vanishes? Generally, it's best to let the user have as much time as they need before the notification goes away. In this app, that could mean:

- letting the notification persist until the user attempts to add another item, at which time the existing notification is replaced, or
- adding a button that the user can click in order to dismiss the message at will, or
- some combination of these two

Right now, the first implementation would be the easiest. You'd only need to delete a little bit of code, if I understand the current code correctly.

**Review by mxmason — APPROVED** [review]
The feature behaves as expected, with a bonus feature to boot! Great work. I've left some notes for an optional UX improvement. Feel free to either merge this branch as-is, or take a shot at my proposed change and request another review! You can also save my suggestion for later, when you'll have time to do some refactoring.

File: src/views/AddItem/AddItem.jsx [hunk 3]
```diff
@@ -19,7 +32,48 @@ export default function AddItem({ token }) {
 
   const handleSubmit = (e) => {
     e.preventDefault();
-    addItem();
+    const checkForErrors = validate();
```
Comment (lindseyindev) [inline src/views/AddItem/AddItem.jsx]: I like the separation here it makes it really easy to move around the validate function

File: src/views/AddItem/AddItem.jsx [hunk 4]
```diff
@@ -19,7 +32,48 @@ export default function AddItem({ token }) {
...
+        frequency: '7',
+      });
+      setError('Item added successfully!');
+      setTimeout(() => {
+        setError('');
+      }, 3000);
+    }
+  };
+
+  const validate = () => {
```
Comment (lindseyindev) [inline src/views/AddItem/AddItem.jsx]: This function works really well and from your PR documentation, you really have really thought about the intended functionality and user experience when going through this interaction!

**Review by lindseyindev — APPROVED** [review]
Amazing work your PR documentation is so thorough it really helped me understand how to test the code! EJ made a great suggestion to move away from the error message time out and you could have a button that a user can use to manually clear the message instead and or allow the message to persist until an item is successfully added.

 One thing I would suggest is thinking about variable and function names which is a way to make code cleaner -- easier & faster to read and understand and now is a great time to try to do more of this.

For example the validate() function - if someone is unfamiliar with this project and the issue was reading that they would have less context for what that function does compared to if it was named according to what that function does or what that function returns
ex: checkForDuplicateItemName or isItemDuplicate (not perfect just an example)

I'm approving this issue, if you have the time to make these changes go for it but if you don't have the availability you can also add them as an issue (document the problem and if you have any suggestions for the solution) as we will have time to come back to those things during the last 2 weeks of the project!

**Pull request merged**